File: src/snapshot.rs
//! Snapshot: walk the org's tables parents-first, scoping each to the org. Collects the in-scope
//! keys (which feed child queries and, later, the stream's initial membership) and streams each
//! table's in-scope rows from source to sink via client-mediated COPY.

use std::collections::{HashMap, HashSet};

use futures_util::{SinkExt, TryStreamExt, pin_mut};
use petgraph::algo::toposort;
use petgraph::graphmap::DiGraphMap;
use serde::Deserialize;
use tokio_postgres::{Client, IsolationLevel, Transaction};

use crate::stream::Membership;

#[derive(Deserialize)]
pub struct Config {
    pub root: String,
    pub relationships: HashMap<String, HashMap<String, Ref>>,
}

#[derive(Deserialize)]
pub struct Ref {
    pub parent: Option<String>, // None for non-FK columns (e.g. file pointers); the walk skips them
    #[serde(default)]
    pub nullable: bool,
}

struct Edge {
    column: String,
    parent: String,
    nullable: bool,
}

struct Graph {
    root: String,
    edges: HashMap<String, Vec<Edge>>, // table -> parent edges
}

impl Graph {
    fn from_config(cfg: &Config) -> Self {
        let mut edges: HashMap<String, Vec<Edge>> = HashMap::new();
        for (table, cols) in &cfg.relationships {
            for (column, r) in cols {
                let Some(parent) = &r.parent else { continue }; // skip non-FK columns (blob pointers)
                edges.entry(table.clone()).or_default().push(Edge {
                    column: column.clone(),
                    parent: parent.clone(),
                    nullable: r.nullable,
                });
            }
        }
        Graph { root: cfg.root.clone(), edges }
    }

    /// Tables in dependency order, root first: every table follows the tables it references.
    /// DiGraphMap is just for test. It only supports one edge per (parent, child) pair.
    /// Need to switch to a real multigraph.
    fn topological_sort(&self) -> Result<Vec<String>, Box<dyn std::error::Error>> {
        let mut g: DiGraphMap<&str, ()> = DiGraphMap::new();
        g.add_node(self.root.as_str());
        for (table, edges) in &self.edges {
            for e in edges {
                g.add_edge(e.parent.as_str(), table.as_str(), ());
            }
        }
        let order = toposort(&g, None).map_err(|c| format!("cycle in table graph at \"{}\"", c.node_id()))?;
        Ok(order.into_iter().map(str::to_string).collect())
    }
}

/// The WHERE predicate scoping <table> to the org: `id = <root_id>` for the root, otherwise
/// `<col> IN (<parent keys>)` for any non-nullable edge -- such a column is on every row, so one
/// edge fully scopes the table; nullable edges are skipped. Returns None if the table has no
/// non-nullable edge, or that parent has no in-scope rows. Reused for both the id select (child
/// scoping) and the COPY extract.
/// Literal IN suits the toy data; a high-cardinality parent is where = ANY($array) would kick in.
fn scope_predicate(graph: &Graph, table: &str, keys: &HashMap<String, Vec<i64>>, root_id: i64) -> Option<(String, String)> {
    if table == graph.root {
        return Some(("root".into(), format!("id = {root_id}")));
    }
    let edge = graph.edges.get(table)?.iter().find(|e| !e.nullable)?;
    let parent_keys = keys.get(&edge.parent).filter(|ks| !ks.is_empty())?;
    let list = parent_keys.iter().map(i64::to_string).collect::<Vec<_>>().join(", ");
    Some((edge.column.clone(), format!("\"{}\" IN ({list})", edge.column)))
}

/// Stream one table's scoped rows source -> sink over the two connections: COPY TO STDOUT frames
/// forwarded chunk-by-chunk into COPY FROM STDIN, so rows are never materialized here. Returns the
/// row count.
async fn copy_table(src: &Transaction<'_>, dst: &Transaction<'_>, table: &str, pred: &str) -> Result<u64, Box<dyn std::error::Error>> {
    let out = src.copy_out(&format!("COPY (SELECT * FROM \"{table}\" WHERE {pred}) TO STDOUT")).await?;
    let sink = dst.copy_in(&format!("COPY \"{table}\" FROM STDIN")).await?;
    pin_mut!(out, sink);
    while let Some(buf) = out.try_next().await? {
        sink.send(buf).await?;
    }
    Ok(sink.finish().await?)
}

/// Run the snapshot: parents-first scoped queries, collecting in-scope keys per table, then copy
/// each table's rows to the sink. All source reads (id selects and COPYs) run in one REPEATABLE
/// READ transaction, so every table is read as of the same frozen snapshot. All sink writes run in
/// one transaction too: the org appears there atomically or not at all.
///
/// Returns the in-scope keys per table -- the stream's initial membership. The caller persists it:
/// membership must reflect what the snapshot saw (i.e. what the sink holds), not the source's
/// later state -- re-deriving it at stream start would silently drop deletes of rows that
/// vanished in between.
pub async fn run_snapshot(source: &mut Client, sink: &mut Client, cfg: &Config, root_id: i64) -> Result<Membership, Box<dyn std::error::Error>> {
    println!("snapshot: scoping org {root_id}\n");
    let graph = Graph::from_config(cfg);
    let tx = source.build_transaction().isolation_level(IsolationLevel::RepeatableRead).start().await?;
    let sink_tx = sink.transaction().await?;

    let mut keys: HashMap<String, Vec<i64>> = HashMap::new();
    let mut scoped: Vec<(String, String, String)> = Vec::new(); // (table, scoped_by, pred) in copy order
    for table in graph.topological_sort()? {
        let Some((scoped_by, pred)) = scope_predicate(&graph, &table, &keys, root_id) else {
            println!("  {table:<16} (no rows in scope)");
            continue;
        };
        // ids feed child scoping (and become the stream's initial membership)
        let ids: Vec<i64> = tx
            .query(&format!("SELECT id FROM \"{table}\" WHERE {pred}"), &[])
            .await?
            .iter()
            .map(|r| r.get::<_, i64>(0))
            .collect();
        keys.insert(table.clone(), ids);
        scoped.push((table, scoped_by, pred));
    }

    // Clear any prior copy of the org from the sink, children first, so re-running is safe
    for (table, _, pred) in scoped.iter().rev() {
        sink_tx.execute(&format!("DELETE FROM \"{table}\" WHERE {pred}"), &[]).await?;
    }

    for (table, scoped_by, pred) in &scoped {
        let copied = copy_table(&tx, &sink_tx, table, pred).await?;
        println!("  {table:<16} via {scoped_by:<18} {copied} row(s) -> sink");
    }

    sink_tx.commit().await?;
    tx.commit().await?; // read-only; just releases the snapshot

    // Membership keeps only tables something references as a parent
    let parents: HashSet<&String> =
        cfg.relationships.values().flat_map(|cols| cols.values()).filter_map(|r| r.parent.as_ref()).collect();
    keys.retain(|table, _| parents.contains(table));
    Ok(keys.into_iter().map(|(table, ids)| (table, ids.into_iter().collect())).collect())
}

File: monarch/snapshot.py
from contextlib import ExitStack
from dataclasses import dataclass

from psycopg import Connection, IsolationLevel, sql

from .config import Cell, Graph
from .utils import trust_sql
from .membership import BlobMembership, Membership


@dataclass
class Source:
    """One store's pinned read (the store is the mover unit; colocated stores hold separate
    connections to their shared database, each on its own exported snapshot). The connection
    must stay inside the slot guard (cli.py)."""

    store: str
    conn: Connection
    snapshot: str  # the slot's exported snapshot name


def read_frozen_ids(
    graph: Graph, source: Cell, conns: dict[str, Connection], org_id: int
) -> dict[str, list[int]]:
    """Each frozen table's ids for the org, read before slot creation: the freeze makes a
    pre-slot read equal the snapshot's view, which is what makes IN-list row filters sound.
    Also seeds a per-store worker's static spine (run_snapshot's static_keys)."""
    out: dict[str, list[int]] = {}
    for table in graph.frozen:
        edge = graph.publication_edge(table)
        if edge is None or edge.parent != graph.root:
            continue
        conn = conns[source.dsn_for(graph.store_of[table])]
        key = graph.primary_key_of[table][0]  # frozen tables are parents: single-key
        rows = conn.execute(
            trust_sql(f'SELECT {key} FROM "{table}" WHERE {edge.column} = %s'), (org_id,)
        ).fetchall()
        out[table] = [r[0] for r in rows]
    return out


def estimate_predicate(
    graph: Graph, table: str, org_id: int, frozen_ids: dict[str, list[int]]
) -> str | None:
    """For ESTIMATION only (EXPLAIN, never executed): exact anchors at root/frozen ids,
    nested semi-joins for dynamic chains so the planner estimates the fanout upfront --
    estimating needs no parent ids, only statistics. Dynamic chains never cross databases
    (config.validate forbids cross-store dynamic scope edges), so the subquery always runs.
    None = not org-scoped; the walk doesn't copy it either."""
    if table == graph.root:
        return f"{graph.primary_key_of[table][0]} = {org_id}"
    if (edge := graph.scope_edge(table)) is None:
        return None
    if edge.parent == graph.root:
        return f"{edge.column} = {org_id}"
    if edge.parent in graph.frozen:
        ids = frozen_ids[edge.parent]
        return f"{edge.column} IN ({', '.join(map(str, ids))})" if ids else "false"
    inner = estimate_predicate(graph, edge.parent, org_id, frozen_ids)
    parent_key = graph.primary_key_of[edge.parent][0]
    return f'{edge.column} IN (SELECT {parent_key} FROM "{edge.parent}" WHERE {inner})'


def estimate_rows(
    conn: Connection,
    graph: Graph,
    tables: list[str],
    org_id: int,
    frozen_ids: dict[str, list[int]],
) -> int:
    """Planner-estimated org rows across `tables`: milliseconds regardless of data size --
    never count(*), which would re-pay the copy's own scan. Complete but inexact; written
    once as copy_rows_estimate (display only, nothing gates on it)."""
    total = 0
    for table in tables:
        if (predicate := estimate_predicate(graph, table, org_id, frozen_ids)) is None:
            continue
        row = conn.execute(
            trust_sql(f'EXPLAIN (FORMAT JSON) SELECT 1 FROM "{table}" WHERE {predicate}')
        ).fetchone()
        assert row is not None
        total += int(row[0][0]["Plan"]["Plan Rows"])
    return total


def scope_predicate(
    graph: Graph, table: str, keys: dict[str, list[int]], root_id: int
) -> tuple[str, sql.Composable] | None:
    """The WHERE predicate scoping <table> to the org: `<primary key> = <root_id>` for the root,
    otherwise `<col> IN (<parent keys>)` for the table's scope edge (graph.scope_edge). Returns
    None if the table has no such edge, or that parent has no in-scope rows. Reused for both the
    key select (child scoping) and the COPY extract.
    Literal IN suits the toy data; a high-cardinality parent is where = ANY(array) would kick in."""
    if table == graph.root:
        return "root", sql.SQL("{} = {}").format(
            sql.Identifier(graph.primary_key_of[table][0]), sql.Literal(root_id)
        )
    edge = graph.scope_edge(table)
    if edge is None:
        return None
    parent_keys = keys.get(edge.parent)
    if not parent_keys:
        return None
    id_list = sql.SQL(", ").join(sql.Literal(k) for k in parent_keys)
    return edge.column, sql.SQL("{} IN ({})").format(sql.Identifier(edge.column), id_list)


def derive_membership(
    sinks: dict[str, Connection], sink: Cell, graph: Graph, root_id: int
) -> Membership:
    """The streams' initial membership, read back from the sink. The sink holds exactly
    the applied-and-acked rows -- snapshot's copy plus streamed changes -- so first start
    and restart are the same read, and a parent grown mid-stream survives a restart.
    Same parents-first walk as the snapshot; parents only, since only they scope others."""
    conn_for = {t: sinks[db.primary_dsn] for db in sink.databases for t in db.tables(graph)}
    keys: dict[str, list[int]] = {}
    for table in graph.topological_sort():
        if table not in graph.parents:
            continue
        if (scope := scope_predicate(graph, table, keys, root_id)) is None:
            continue
        _, pred = scope
        select = sql.SQL("SELECT {} FROM {} WHERE {}").format(
            sql.Identifier(graph.primary_key_of[table][0]), sql.Identifier(table), pred
        )
        keys[table] = [r[0] for r in conn_for[table].execute(select).fetchall()]
    return {table: set(ids) for table, ids in keys.items()}


def copy_table(source: Connection, sink: Connection, table: str, pred: sql.Composable) -> int:
    """Stream one table's scoped rows source -> sink: COPY TO STDOUT frames forwarded
    chunk-by-chunk into COPY FROM STDIN, so rows are never materialized here."""
    with source.cursor() as out_cur, sink.cursor() as in_cur:
        out = sql.SQL("COPY (SELECT * FROM {} WHERE {}) TO STDOUT").format(
            sql.Identifier(table), pred
        )
        into = sql.SQL("COPY {} FROM STDIN").format(sql.Identifier(table))
        with out_cur.copy(out) as reader, in_cur.copy(into) as writer:
            for data in reader:
                writer.write(data)
        return in_cur.rowcount  # set once the copy block closes; cleared when the cursor closes


def record_scoped_keys(
    source: Connection,
    table: str,
    pred: sql.Composable,
    graph: Graph,
    blob_members: dict[str, BlobMembership],
) -> int:
    """Record the blob keys behind <table>'s blob columns for in-scope rows. No bytes move
    here: the copy worker converges membership into the sink bucket, and blob-before-row
    binds only at cut-over (the staging sink serves no reads). Keys are read in the pinned
    transaction."""
    recorded = 0
    for column, store in graph.blobs.get(table, {}).items():
        keys = sql.SQL("SELECT DISTINCT {c} FROM {t} WHERE {p} AND {c} IS NOT NULL").format(
            c=sql.Identifier(column), t=sql.Identifier(table), p=pred
        )
        for (key,) in source.execute(keys).fetchall():
            blob_members[store].add(key)
            recorded += 1
    return recorded


def run_snapshot(
    sources: list[Source],
    sinks: dict[str, Connection],
    sink: Cell,
    graph: Graph,
    root_id: int,
    blob_members: dict[str, BlobMembership],
    static_keys: dict[str, list[int]] | None = None,
) -> tuple[Membership, dict[str, int]]:
    """Run the snapshot across the passed stores: parents-first scoped queries collecting
    in-scope keys per table -- the keys dict is shared, so a table's predicate can consume
    parent keys read from another store -- then copy each table's rows to its sink database.

    static_keys pre-seeds the keys dict with the static spine (root + static tables) so a
    single-store call can scope its tables against parents that live in other stores, handed
    in rather than read during the walk -- sound because those ids are static for the move
    (config.validate confines cross-store scope edges to them). None when `sources` already
    spans every store (the CLI's one-shot snapshot), so every parent is present to read
    directly. Either way the copy is only the org's scoped rows -- static_keys sets how many
    stores a call spans, never what's in scope.

    Each store is read in one REPEATABLE READ transaction pinned to its own slot's
    exported snapshot: consistent per store. The stores' consistent points differ
    slightly (even colocated ones -- separate slots); each store's stream resumes exactly
    at its own, so nothing is missed; cross-store edges stay safe because they bind only
    the static spine. Sink
    writes are one transaction per sink database -- atomic per database, not across them
    (that would need 2PC).

    Returns (membership, copied): membership is the in-scope parent-table ids, kept for the
    in-process stream handoff (the CLI streams re-derive from the sink instead -- deriving
    from the *source* would silently drop deletes of rows that vanished between snapshot and
    stream start). copied is the actual rows written per table -- the accurate copy total,
    since leaf tables never appear in membership."""
    print(f"snapshot: scoping org {root_id}\n")
    my_tables = {t for s in sources for t in graph.store_tables(s.store)}
    source_for = {t: s.conn for s in sources for t in graph.store_tables(s.store)}
    sink_for = {
        t: sinks[db.primary_dsn]
        for db in sink.databases
        for t in db.tables(graph)
        if t in my_tables
    }
    for s in sources:
        s.conn.isolation_level = IsolationLevel.REPEATABLE_READ
    with ExitStack() as stack:
        for s in sources:
            stack.enter_context(s.conn.transaction())
            # SET TRANSACTION SNAPSHOT must run before the transaction's first query (else the
            # transaction already has its own snapshot) and requires REPEATABLE READ.
            s.conn.execute(sql.SQL("SET TRANSACTION SNAPSHOT {}").format(sql.Literal(s.snapshot)))
        for conn in sinks.values():
            stack.enter_context(conn.transaction())

        keys: dict[str, list[int]] = dict(static_keys or {})
        scoped: list[tuple[str, str, sql.Composable]] = []  # (table, scoped_by, pred) in copy order
        for table in graph.topological_sort():
            if table not in my_tables:  # another store's territory: its own worker copies it
                continue
            if (scope := scope_predicate(graph, table, keys, root_id)) is None:
                print(f"  {table:<16} (no rows in scope)")
                continue
            scoped_by, pred = scope
            if table in graph.parents:
                # keys feed child scoping (and become the streams' initial membership);
                # only parents scope others, and only they are single-key (config.validate)
                select = sql.SQL("SELECT {} FROM {} WHERE {}").format(
                    sql.Identifier(graph.primary_key_of[table][0]), sql.Identifier(table), pred
                )
                keys[table] = [r[0] for r in source_for[table].execute(select).fetchall()]
            scoped.append((table, scoped_by, pred))

        # Clear any prior copy of the org from the sink, children first, so re-running is safe
        for table, _, pred in reversed(scoped):
            sink_for[table].execute(
                sql.SQL("DELETE FROM {} WHERE {}").format(sql.Identifier(table), pred)
            )

        copied_by_table: dict[str, int] = {}
        for table, scoped_by, pred in scoped:
            blobs = record_scoped_keys(source_for[table], table, pred, graph, blob_members)
            copied = copy_table(source_for[table], sink_for[table], table, pred)
            copied_by_table[table] = copied
            extra = f" + {blobs} key(s)" if table in graph.blobs else ""
            print(f"  {table:<16} via {scoped_by:<18} {copied} row(s){extra} -> sink")

    # keys holds only parent tables; copied_by_table has the row count for every copied table
    return {table: set(ids) for table, ids in keys.items()}, copied_by_table
